### PR TITLE
Add support for hints' based autodiscovery in kubernetes provider

### PR DESCRIPTION
**Pull request by ChrsMark**
## What does this PR do?

Elastic Agent implementation to support hints based autodiscovery. See https://github.com/elastic/elastic-agent/issues/662 and  https://github.com/elastic/elastic-agent/issues/613 for more details 

## Why is it important?

In order to support https://github.com/elastic/elastic-agent/issues/613

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [x] Feature Implementation
- [x] Add new unit tests for the new functionality

## How to test this PR locally

1. Deploy a target redis pod:

```yml
apiVersion: v1
kind: Pod
metadata:
  name: redis
  annotations:
    co.elastic.hints/package: redis
    co.elastic.hints/data_streams: info, key
    co.elastic.hints/host: '${kubernetes.pod.ip}:6379'
    co.elastic.hints/info.period: 1m
    co.elastic.hints/key.period: 10m
  labels:
    k8s-app: redis
    app: redis
spec:
  containers:
  - image: redis
    imagePullPolicy: IfNotPresent
    name: redis
    ports:
    - name: redis
      containerPort: 6379
      protocol: TCP

```

2. Configure Agent properly

<details>

<summary>elastic-agent.yml</summary>

```yml
providers:
  kubernetes:
    kube_config: /home/chrismark/.kube/config
    node: "kind-control-plane"
    hints.enabled: true
inputs:
  - name: filestream-redis
    type: filestream
    use_output: default
    streams:
      - condition: ${kubernetes.hints.redis.log.enabled} == true or ${kubernetes.hints.redis.enabled} == true
        data_stream:
          dataset: redis.log
          type: logs
        exclude_files:
          - .gz$
        exclude_lines:
          - ^\s+[\-`('.|_]
        parsers:
          - container:
              format: auto
              stream: ${kubernetes.hints.redis.log.stream|'all'}
        paths:
          - /var/log/containers/*${kubernetes.hints.container_id}.log
        prospector:
          scanner:
            symlinks: true
        tags:
          - redis-log
    data_stream.namespace: default
  - name: redis-redis
    type: redis
    use_output: default
    streams:
      - condition: ${kubernetes.hints.redis.slowlog.enabled} == true or ${kubernetes.hints.redis.enabled} == true
        data_stream:
          dataset: redis.slowlog
          type: logs
        hosts:
          - ${kubernetes.hints.redis.slowlog.host|'127.0.0.1:6379'}
        password: ${kubernetes.hints.redis.slowlog.password|''}
    data_stream.namespace: default
  - name: redis/metrics-redis
    type: redis/metrics
    use_output: default
    streams:
      - condition: ${kubernetes.hints.redis.info.enabled} == true or ${kubernetes.hints.redis.enabled} == true
        data_stream:
          dataset: redis.info
          type: metrics
        hosts:
          - ${kubernetes.hints.redis.info.host|'127.0.0.1:6379'}
        idle_timeout: 20s
        maxconn: 10
        metricsets:
          - info
        network: tcp
        password: ${kubernetes.hints.redis.info.password|''}
        period: ${kubernetes.hints.redis.info.period|'10s'}
      - condition: ${kubernetes.hints.redis.key.enabled} == true or ${kubernetes.hints.redis.enabled} == true
        data_stream:
          dataset: redis.key
          type: metrics
        hosts:
          - ${kubernetes.hints.redis.key.host|'127.0.0.1:6379'}
        idle_timeout: 20s
        key.patterns:
          - limit: 20
            pattern: '*'
        maxconn: 10
        metricsets:
          - key
        network: tcp
        password: ${kubernetes.hints.redis.key.password|''}
        period: ${kubernetes.hints.redis.key.period|'10s'}
      - condition: ${kubernetes.hints.redis.keyspace.enabled} == true or ${kubernetes.hints.redis.enabled} == true
        data_stream:
          dataset: redis.keyspace
          type: metrics
        hosts:
          - ${kubernetes.hints.redis.keyspace.host|'127.0.0.1:6379'}
        idle_timeout: 20s
        maxconn: 10
        metricsets:
          - keyspace
        network: tcp
        password: ${kubernetes.hints.redis.keyspace.password|''}
        period: ${kubernetes.hints.redis.keyspace.period|'10s'}
    data_stream.namespace: default
```

</details>

3. Run inspect command and verify that the configuration is populated properly for the redis Pod
`./elastic-agent inspect output -o default -p -v -d "*" -c elastic-agent.yml `

<details>

<summary>Example output</summary>

```console
[default] filebeat:
filebeat:
  inputs:
  - hosts:
    - 127.0.0.1:6379
    index: logs-redis.slowlog-default
    name: redis-redis
    password: ""
    processors:
    - add_fields:
        fields:
          dataset: redis.slowlog
          namespace: default
          type: logs
        target: data_stream
    - add_fields:
        fields:
          dataset: redis.slowlog
        target: event
    - add_fields:
        fields:
          id: ""
          snapshot: false
          version: 8.5.0
        target: elastic_agent
    - add_fields:
        fields:
          id: ""
        target: agent
    type: redis
  - exclude_files:
    - .gz$
    exclude_lines:
    - ^\s+[\-`('.|_]
    index: logs-redis.log-default
    name: filestream-redis
    parsers:
    - container:
        format: auto
        stream: all
    paths:
    - /var/log/containers/*f5fe0acafa56d55291aec1df632cb33c769c7ad2a1391bfd0abe884f554dfb5e.log
    processors:
    - add_fields:
        fields:
          id: f5fe0acafa56d55291aec1df632cb33c769c7ad2a1391bfd0abe884f554dfb5e
          image:
            name: redis
          runtime: containerd
        target: container
    - add_fields:
        fields:
          container:
            name: redis
          labels:
            app: redis
            k8s-app: redis
          namespace: default
          namespace_labels:
            kubernetes_io/metadata_name: default
          namespace_uid: a22a6d0a-8687-43cd-837c-b567c3e1fac2
          node:
            hostname: kind-control-plane
            labels:
              beta_kubernetes_io/arch: amd64
              beta_kubernetes_io/os: linux
              kubernetes_io/arch: amd64
              kubernetes_io/hostname: kind-control-plane
              kubernetes_io/os: linux
              node-role_kubernetes_io/control-plane: ""
              node_kubernetes_io/exclude-from-external-load-balancers: ""
            name: kind-control-plane
            uid: 42fe39bb-19dd-4382-85be-c8c1d67ab742
          pod:
            ip: 10.244.0.2
            name: redis
            uid: ad62ad95-fa2f-4884-9ba1-7bc34080b860
        target: kubernetes
    - add_fields:
        fields:
          cluster:
            name: kind-kind
            url: https://127.0.0.1:42447
        target: orchestrator
    - add_fields:
        fields:
          dataset: redis.log
          namespace: default
          type: logs
        target: data_stream
    - add_fields:
        fields:
          dataset: redis.log
        target: event
    - add_fields:
        fields:
          id: ""
          snapshot: false
          version: 8.5.0
        target: elastic_agent
    - add_fields:
        fields:
          id: ""
        target: agent
    prospector:
      scanner:
        symlinks: true
    tags:
    - redis-log
    type: filestream
output:
  elasticsearch:
    api-key: example-key
    hosts:
    - 127.0.0.1:9200
---[default] metricbeat:
metricbeat:
  modules:
  - hosts:
    - 10.244.0.2:6379
    idle_timeout: 20s
    index: metrics-redis.info-default
    maxconn: 10
    metricsets:
    - info
    module: redis
    name: redis/metrics-redis
    network: tcp
    password: ""
    period: 1m
    processors:
    - add_fields:
        fields:
          labels:
            app: redis
            k8s-app: redis
          namespace: default
          namespace_labels:
            kubernetes_io/metadata_name: default
          namespace_uid: a22a6d0a-8687-43cd-837c-b567c3e1fac2
          node:
            hostname: kind-control-plane
            labels:
              beta_kubernetes_io/arch: amd64
              beta_kubernetes_io/os: linux
              kubernetes_io/arch: amd64
              kubernetes_io/hostname: kind-control-plane
              kubernetes_io/os: linux
              node-role_kubernetes_io/control-plane: ""
              node_kubernetes_io/exclude-from-external-load-balancers: ""
            name: kind-control-plane
            uid: 42fe39bb-19dd-4382-85be-c8c1d67ab742
          pod:
            ip: 10.244.0.2
            name: redis
            uid: ad62ad95-fa2f-4884-9ba1-7bc34080b860
        target: kubernetes
    - add_fields:
        fields:
          cluster:
            name: kind-kind
            url: https://127.0.0.1:42447
        target: orchestrator
    - add_fields:
        fields:
          dataset: redis.info
          namespace: default
          type: metrics
        target: data_stream
    - add_fields:
        fields:
          dataset: redis.info
        target: event
    - add_fields:
        fields:
          id: ""
          snapshot: false
          version: 8.5.0
        target: elastic_agent
    - add_fields:
        fields:
          id: ""
        target: agent
  - hosts:
    - 10.244.0.2:6379
    idle_timeout: 20s
    index: metrics-redis.key-default
    key.patterns:
    - limit: 20
      pattern: '*'
    maxconn: 10
    metricsets:
    - key
    module: redis
    name: redis/metrics-redis
    network: tcp
    password: ""
    period: 10m
    processors:
    - add_fields:
        fields:
          labels:
            app: redis
            k8s-app: redis
          namespace: default
          namespace_labels:
            kubernetes_io/metadata_name: default
          namespace_uid: a22a6d0a-8687-43cd-837c-b567c3e1fac2
          node:
            hostname: kind-control-plane
            labels:
              beta_kubernetes_io/arch: amd64
              beta_kubernetes_io/os: linux
              kubernetes_io/arch: amd64
              kubernetes_io/hostname: kind-control-plane
              kubernetes_io/os: linux
              node-role_kubernetes_io/control-plane: ""
              node_kubernetes_io/exclude-from-external-load-balancers: ""
            name: kind-control-plane
            uid: 42fe39bb-19dd-4382-85be-c8c1d67ab742
          pod:
            ip: 10.244.0.2
            name: redis
            uid: ad62ad95-fa2f-4884-9ba1-7bc34080b860
        target: kubernetes
    - add_fields:
        fields:
          cluster:
            name: kind-kind
            url: https://127.0.0.1:42447
        target: orchestrator
    - add_fields:
        fields:
          dataset: redis.key
          namespace: default
          type: metrics
        target: data_stream
    - add_fields:
        fields:
          dataset: redis.key
        target: event
    - add_fields:
        fields:
          id: ""
          snapshot: false
          version: 8.5.0
        target: elastic_agent
    - add_fields:
        fields:
          id: ""
        target: agent
  - hosts:
    - 127.0.0.1:6379
    idle_timeout: 20s
    index: metrics-redis.keyspace-default
    maxconn: 10
    metricsets:
    - keyspace
    module: redis
    name: redis/metrics-redis
    network: tcp
    password: ""
    period: 10s
    processors:
    - add_fields:
        fields:
          labels:
            app: redis
            k8s-app: redis
          namespace: default
          namespace_labels:
            kubernetes_io/metadata_name: default
          namespace_uid: a22a6d0a-8687-43cd-837c-b567c3e1fac2
          node:
            hostname: kind-control-plane
            labels:
              beta_kubernetes_io/arch: amd64
              beta_kubernetes_io/os: linux
              kubernetes_io/arch: amd64
              kubernetes_io/hostname: kind-control-plane
              kubernetes_io/os: linux
              node-role_kubernetes_io/control-plane: ""
              node_kubernetes_io/exclude-from-external-load-balancers: ""
            name: kind-control-plane
            uid: 42fe39bb-19dd-4382-85be-c8c1d67ab742
          pod:
            ip: 10.244.0.2
            name: redis
            uid: ad62ad95-fa2f-4884-9ba1-7bc34080b860
        target: kubernetes
    - add_fields:
        fields:
          cluster:
            name: kind-kind
            url: https://127.0.0.1:42447
        target: orchestrator
    - add_fields:
        fields:
          dataset: redis.keyspace
          namespace: default
          type: metrics
        target: data_stream
    - add_fields:
        fields:
          dataset: redis.keyspace
        target: event
    - add_fields:
        fields:
          id: ""
          snapshot: false
          version: 8.5.0
        target: elastic_agent
    - add_fields:
        fields:
          id: ""
        target: agent
output:
  elasticsearch:
    api-key: example-key
    hosts:
    - 127.0.0.1:9200
---[default] FLEET_MONITORING:
monitoring_checksum: d4f14553cdd37ff68bb492de0eea7256
output:
  elasticsearch:
    api-key: example-key
    hosts:
    - 127.0.0.1:9200
    type: elasticsearch
programs:
- filebeat
- metricbeat
---% 
```

</details>

Extra:

##### Test with various container logs' streams

1. Deploy apache
```yml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: httpd-deployment-xfusion
  annotations:
    co.elastic.hints/package: apache
    co.elastic.hints/data_streams: access, error
    #co.elastic.hints/access.stream: stdout
    #co.elastic.hints/error.stream: stderr
  labels:
    app: httpd_app_xfusion
spec :
  replicas: 1
  selector:
    matchLabels:
      app: httpd_app_xfusion
  template:
    metadata:
      annotations:
        co.elastic.hints/package: apache
        co.elastic.hints/data_streams: access, error
        co.elastic.hints/access.stream: stdout
        co.elastic.hints/error.stream: stderr
      labels:
        hints: parse
        app: httpd_app_xfusion
    spec :
      containers:
        - name: httpd-container-xfusion
          image: httpd:latest
          ports:
            - containerPort : 80
---
apiVersion: v1
kind: Service
metadata:
  name: httpd-service-xfusion
spec:
  type: ClusterIP
  selector:
    app: httpd_app_xfusion
  ports:
    - port: 8086
      targetPort: 80
```

2.

<details>

<summary>Template for apache</summary>

```yml
inputs:
  - name: filestream-apache
    type: filestream
    use_output: default
    streams:
      - condition: ${kubernetes.hints.apache.access.enabled} == true or ${kubernetes.hints.apache.enabled} == true
        data_stream:
          dataset: apache.access
          type: logs
        exclude_files:
          - .gz$
        parsers:
          - container:
              format: auto
              stream: ${kubernetes.hints.apache.access.stream|'all'}
        paths:
          - /var/log/containers/*${kubernetes.hints.container_id}.log
        prospector:
          scanner:
            symlinks: true
        tags:
          - apache-access
      - condition: ${kubernetes.hints.apache.error.enabled} == true or ${kubernetes.hints.apache.enabled} == true
        data_stream:
          dataset: apache.error
          type: logs
        exclude_files:
          - .gz$
        parsers:
          - container:
              format: auto
              stream: ${kubernetes.hints.apache.error.stream|'all'}
        paths:
          - /var/log/containers/*${kubernetes.hints.container_id}.log
        processors:
          - add_locale: null
        prospector:
          scanner:
            symlinks: true
        tags:
          - apache-error
    data_stream.namespace: default
```

</details>

3. Verify that streams are assigned properly per data_stream like:
```
...
name: filestream-apache
    parsers:
    - container:
        format: auto
        stream: stderr
    paths:
    - /var/log/containers/*a5289bb2da8960516a5b3d5c3a484ececc349e7c9a2118209f3c145f73e7cbb3.log
...
name: filestream-apache
    parsers:
    - container:
        format: auto
        stream: stdout
    paths:
    - /var/log/containers/*a5289bb2da8960516a5b3d5c3a484ececc349e7c9a2118209f3c145f73e7cbb3.log
```
##### Default container stream

<details>

<summary> prometheus.yml</summary>

```yml
inputs:
  - name: filestream-prometheus
    type: filestream
    use_output: default
    streams:
      - condition: ${kubernetes.hints.prometheus.container_logs.enabled} == true
        data_stream:
          dataset: kubernetes.container_logs
          type: logs
        exclude_files: []
        exclude_lines: []
        parsers:
          - container:
              format: auto
              stream: all
        paths:
          - /var/log/containers/*${kubernetes.hints.container_id}.log
        prospector:
          scanner:
            symlinks: true
        tags: []
    data_stream.namespace: default
  - name: prometheus/metrics-prometheus
    type: prometheus/metrics
    use_output: default
    streams:
      - bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
        condition: ${kubernetes.hints.prometheus.collector.enabled} == true or ${kubernetes.hints.prometheus.enabled} == true
        data_stream:
          dataset: prometheus.collector
          type: metrics
        hosts:
          - ${kubernetes.hints.prometheus.collector.host|'localhost:9090'}
        metrics_filters.exclude: null
        metrics_filters.include: null
        metrics_path: /metrics
        metricsets:
          - collector
        password: ${kubernetes.hints.prometheus.collector.password|'secret'}
        period: ${kubernetes.hints.prometheus.collector.period|'10s'}
        rate_counters: true
        ssl.certificate_authorities:
          - /var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt
        use_types: true
        username: ${kubernetes.hints.prometheus.collector.username|'user'}
```

</details>

## Related issues

Implements part 1 of https://github.com/elastic/elastic-agent/issues/662

### CODE REVIEW

**Comment by ChrsMark** [comment]
Blocked by https://github.com/elastic/elastic-agent/issues/1107

**Comment by cmacknz** [comment]
@blakerouse does this interact with the changes on the [feature-arch-v2](https://github.com/elastic/elastic-agent/tree/feature-arch-v2) branch at all? I don't think providers changed at all but I want to double check.

@ChrsMark FYI we have re-implemented large portions of the agent internals related to starting and controlling processes on a separate feature branch while we migrate the affected inputs/beats to the new APIs. My hope is this change is mostly unaffected by what is on that branch assuming you want this merged for 8.5.

**Comment by ChrsMark** [comment]
> @blakerouse does this interact with the changes on the [feature-arch-v2](https://github.com/elastic/elastic-agent/tree/feature-arch-v2) branch at all? I don't think providers changed at all but I want to double check.
> 
> @ChrsMark FYI we have re-implemented large portions of the agent internals related to starting and controlling processes on a separate feature branch while we migrate the affected inputs/beats to the new APIs. My hope is this change is mostly unaffected by what is on that branch assuming you want this merged for 8.5.

Thanks @cmacknz for raising this. This PR only extends the existent provider's mechanism so I hope we are safe here. Let's wait for @blakerouse to confirm though:) .

File: internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ type contextProviderK8sSecrets struct {
 }
 
 // ContextProviderBuilder builds the context provider.
```
Comment (gizas) [inline internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go:36]: Put a comment here what the managed boolean flag provides

File: internal/pkg/composable/providers/kubernetes/hints.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+func (m *hintsBuilder) getStreamPassword(hints mapstr.M, streamName string) string {
+	key := fmt.Sprintf("%v.%v", streamName, password)
+	return utils.GetHintString(hints, m.Key, key)
+}
+
+func (m *hintsBuilder) getFromMeta(value string, kubeMeta mapstr.M) string {
+	if value == "" {
+		return ""
+	}
+	r := regexp.MustCompile(`\${([^{}]+)}`)
```
Comment (gizas) [inline internal/pkg/composable/providers/kubernetes/hints.go:102]: Why is this used?

File: internal/pkg/composable/providers/kubernetes/pod_test.go [hunk 3]
```diff
@@ -274,13 +293,21 @@ func TestEphemeralContainers(t *testing.T) {
 		context.TODO(),
 		providerDataChan,
 	}
+
+	logger := getLogger()
+	var cfg Config
+	c := config.New()
+	_ = c.Unpack(&cfg)
 	generateContainerData(
```
Comment (gizas) [inline internal/pkg/composable/providers/kubernetes/pod_test.go:301]: Do you like it like this? Maybe inlined?

**Review by gizas — APPROVED** [review]

File: internal/pkg/composable/providers/kubernetes/hints.go (hunk 2 above)
Comment (ChrsMark) [reply internal/pkg/composable/providers/kubernetes/hints.go:102]: Within the hint's string you might have sth like `${kubernetes.pod.ip}` which you want to be retrieved dynamically from the Pod's meta. So here we have a regexp to extract the actual value of the variable we target out of a `"${kubernetes.*}"` hint.

To be more specific imagine that you have a hint like `'${kubernetes.pod.ip}:6379'`, then you want to replace the `${kubernetes.pod.ip}` part with value from the Pod's metadata so as to have sth like `10.28.90.345:6379`.

Comment (ChrsMark) [reply internal/pkg/composable/providers/kubernetes/hints.go:105]: added this as comment within the code too;) : [8b50e76](https://github.com/elastic/elastic-agent/pull/698/commits/8b50e766ed354692602f37c26a448dcc44abc044)

File: internal/pkg/composable/providers/kubernetes/pod.go [hunk 4]
```diff
@@ -149,18 +154,40 @@ func (p *pod) emitRunning(pod *kubernetes.Pod) {
...
-	// get Pod metadata (not specific to any container)
-	_ = p.comm.AddOrUpdate(data.uid, PodPriority, data.mapping, data.processors)
+
+	if p.config.Hints.Enabled() { // This is "hints based autodiscovery flow"
+		if !p.managed {
+			if ann, ok := data.mapping["annotations"]; ok {
+				annotations, _ := ann.(mapstr.M)
+				hints := utils.GenerateHints(annotations, "", p.config.Prefix)
+				if len(hints) > 0 {
+					p.logger.Errorf("Extracted hints are :%v", hints)
```
Comment (blakerouse) [inline internal/pkg/composable/providers/kubernetes/pod.go:164]: Shouldn't this be debug level?

File: internal/pkg/composable/providers/kubernetes/pod.go [hunk 5]
```diff
@@ -149,18 +154,40 @@ func (p *pod) emitRunning(pod *kubernetes.Pod) {
...
+					hintsMapping := GenerateHintsMapping(hints, data.mapping, p.logger, "")
+					p.logger.Errorf("Generated hints mappings are :%v", hintsMapping)
```
Comment (blakerouse) [inline internal/pkg/composable/providers/kubernetes/pod.go:166]: Shouldn't this be debug level?

**Review by blakerouse — COMMENTED** [review]
This should work with v2 just fine. Nothing changed in the providers or the variable substitution.

**Pull request merged**